scalafix/rules/src/main/scala/rsc/rules/pretty/SemanticdbPrinter.scala

```diff
@@ -1,18 +1,88 @@
 // Copyright (c) 2017-2018 Twitter, Inc.
 // Licensed under the Apache License, Version 2.0 (see LICENSE.md).
 // NOTE: This file has been partially copy/pasted from scalameta/scalameta.
-package scalafix.internal.rule.pretty
+package rsc.rules.pretty
 
 import rsc.lexis._
 import rsc.pretty._
+import rsc.rules.semantics._
 import scala.collection.mutable
+import scala.meta._
 import scala.meta.internal.{semanticdb => s}
 import scala.meta.internal.semanticdb.Scala._
+import scala.meta.internal.semanticdb.Scala.{Descriptor => d}
 import scala.meta.internal.semanticdb.SymbolInformation.{Kind => k}
 import scala.meta.internal.semanticdb.SymbolInformation.{Property => p}
-import scalafix.internal.rule.semantics._
+import scalafix.v0._
+
+class SemanticdbPrinter(env: Env, index: DocumentIndex) extends Printer {
+  def pprint(tree: s.Tree): Unit = tree match {
+    case s.OriginalTree(range) =>
+      str(index.substring(range).get)
+    case s.ApplyTree(fn, args) =>
+      pprint(fn)
+      rep("(", args, ", ", ")")(pprint)
+    case s.TypeApplyTree(fn, targs) =>
+      pprint(fn)
+      rep("[", targs, ", ", "]")(pprint)
+    case s.SelectTree(qual, id) =>
+      // FIXME: https://github.com/twitter/rsc/issues/142
+      val needsParens = qual match {
+        case s.OriginalTree(range) =>
+          val originalTerm = index.substring(range).get.parse[Term].get
+          originalTerm match {
+            case _: Term.ApplyInfix => true
+            case _ => false
+          }
+        case _ => false
+      }
+      if (needsParens) str("(")
+      pprint(qual)
+      if (needsParens) str(")")
+      str(".")
+      pprint(id.get.sym)
+    case s.IdTree(sym) =>
+      sym.owner.desc match {
+        case d.None =>
+          pprint(sym)
+        case _: d.Package | _: d.Term =>
+          pprint(s.SelectTree(s.IdTree(sym.owner), Some(s.IdTree(sym))))
+        case d.Type(name) =>
+          if (env.lookupThis(name) == sym.owner) {
+            pprint(sym.owner)
+            str(".this.")
+            pprint(sym)
+          } else {
+            // TODO: This looks incorrect.
+            // str(".")
+            ???
+          }
+        case desc => sys.error(s"unsupported desc $desc")
+      }
+    case s.FunctionTree(params, term) =>
+      str("{")
+      params match {
+        case Seq() => str("() => ")
+        case Seq(id) =>
+          pprint(id.sym)
+          str(" => ")
+        case _ =>
+          rep("(", params, ", ", ") => ")(id => pprint(id.sym))
+      }
+      pprint(term)
+      str("}")
+    case s.MacroExpansionTree(expandee, _) =>
+      expandee match {
+        case s.ApplyTree(
+            s.IdTree("scala/reflect/package.materializeClassTag()."),
+            Nil) =>
+          str("_root_.scala.reflect.`package`.classTag")
```

Comment:
@maxov This is needed to work around the extravagant way how ClassTag synthesis is done in Scalac: https://github.com/scala/scala/blob/v2.12.6/src/library/scala/reflect/package.scala#L59-L63. Adding this mechanism to Scalac wasn't my brightest idea 😭